### PR TITLE
feat: add Payout Bounty button

### LINKED ISSUE #1: Integrate Bounty Payout form
# Bounty

Integrate with BountyPayout contract same as we have in our old Gnosis Multisig fork.

## Scope

- add a button to start Bounty Payout similar to our Gnosis Multisig
- decode payout transactions on the list showing details in human-readable way
- (optionally) include ABI so that people don't have to add it manually

## Deliverables

- updated code

## Funding Circle

General Circle

## Bounty Owner/Gardener

@troggy as **Process Engineer**

## Gain for the Role

LeapDAO can use existing processes with Gnosis Safe which is superior in terms of speed and gas usage

## Roles
bounty gardener: @troggy / 50 DAI 
bounty worker: @troggy / 670 DAI
bounty reviewer: @Clearwood  /  180 DAI 

## Gardener checklist

- [x] I have attached one of the `size-` labels to this bounty
- [x] I have submitted this bounty via http://bounty.leapdao.org/viewform
- [x] I have added this bounty to bounty board here https://github.com/orgs/leapdao/projects/6

### PR DESCRIPTION
**Pull request by troggy**
Please remember the [Contributing Guidelines](https://github.com/leapdao/meta/blob/master/CONTRIBUTION.md) :heart:

## What does this PR do?

- adds Payout Bounty button to wallet's "Send" modal
- decode and show bounty payout details on "Transactions" screen
- show DAI balance on Send Bounty screen

## Recommendations for testing

Use cases:
- [ ] bounty with gardener, reviewer and worker
- [ ] bounty with gardener and worker
- [ ] bounty with gardener and reviewer
- [ ] bounty with rep only for gardener
- [ ] bounty with rep only for worker
- [ ] bounty with rep only for reviewer
- [ ] validations etc
- [ ] bounty link encoding/decoding
- [ ] address book integration on Create Payout screen
- [ ] address book integration on Review screen
- [ ] address book integration on Transactions screen

## Links to relevant issues or information

Closes #1

### CODE REVIEW

**Comment by troggy** [comment]
Preview image

**Comment by troggy** [comment]
Created Rinkeby Safe for testing: 0xbEcA6fD0f0B1DA5652816d862a3284BB53bCf4cF

**Comment by troggy** [comment]
To run on mainnet register with blocknative.com first

then

```sh
REACT_APP_BLOCKNATIVE_KEY=<Blocknative API key> REACT_APP_NETWORK=mainnet REACT_APP_INFURA_TOKEN=<infura project id> yarn start
```

**Comment by troggy** [comment]
Create Rinkeby Bounty Payout: 0x8F36E3d1235d3CA958Dce31af1afE75E451dFcD2

Uses tLEAP instead of DAI and doesn't call Colony. 

To use it the Safe needs to be whitelisted on it first. Call from LeapDAO Rinkeby Multisig (0x03CAa70D57c9E98c0CeAf1EF95f9C9459B89f1Bf):
```
(await BountyPayout.at("0x8F36E3d1235d3CA958Dce31af1afE75E451dFcD2")).addWhitelisted("<your safe address>")
```

Sample TX: https://rinkeby.etherscan.io/tx/0xe9beaf3c6cd6d3bb4cdc29a8d01116cc31d9778a40b5986663571ee4befe8206

**Comment by troggy** [comment]
Some screens:
image
image

**Comment by Clearwood** [comment]
First of all it looks awesome! :)
Some feedback: 
1. right now the payment screen crashes if the token balance of a given safe is 0.
2. The check for the validity of a payout address does not seem to work, the validity of the link to the GitHub issue however is not checked.
image
3. The environment variables which need to be set should be added to the README e.g. REACT_APP_INFURA_TOKEN
4. Clicking on a failed or pending transaction leads to a crash
image

**Comment by troggy** [comment]
Thanks for the grill! Fixing..

> The environment variables which need to be set should be added to the README e.g. REACT_APP_INFURA_TOKEN

Addressed here #5 in a separate bounty

**Comment by Clearwood** [comment]
The fixes look great :) 
One thing that I could see now is that the url formatting looks a bit off 
image

**Comment by troggy** [comment]
msg data to approve 10000 tLEAP for Rinkeby Bounty payout:

```
0x095ea7b30000000000000000000000008f36e3d1235d3ca958dce31af1afe75e451dfcd200000000000000000000000000000000000000000000021e19e0c9bab2400000
```

**Comment by troggy** [comment]
final result
image

**Comment by Clearwood** [comment]
Great work :) There seems to be a problem with the payout of Decimals now 
image
Do you want to hide the Apps for the time being?

**Comment by troggy** [comment]
oh.. never tested decimals 😅

**Comment by troggy** [comment]
to hide the Apps run with `REACT_APP_ENV=production`. This is probably change already in upstream as Apps are live on gnosis-safe.io

**Comment by troggy** [comment]
@Clearwood solved decimal issue + solved one more issues with boutny link > 32 symbols (bounty link in hex should fit in 32 bytes)

File: src/routes/safe/components/Balances/SendModal/screens/PayoutBounty/index.jsx [hunk 1]
```diff
@@ -146,7 +146,7 @@ const PayoutBounty = ({ initialValues, onClose, onNext }: Props) => {
                       placeholder="Link to bounty ticket"
                       text="Link to bounty ticket"
                       type="text"
-                      validate={required}
+                      validate={composeValidators(required, minMaxLength(1, 32))}
```
Comment (Clearwood) [inline src/routes/safe/components/Balances/SendModal/screens/PayoutBounty/index.jsx]: The length of the string will almost always exceed 32 characters, maybe check the length after removing the `https://github.com/leapdao/` bit.

Comment (troggy) [reply src/routes/safe/components/Balances/SendModal/screens/PayoutBounty/index.jsx]: very good point 🤦‍♂️

File: src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/index.jsx [hunk 2]
```diff
@@ -208,6 +223,57 @@ const CustomDescription = ({ amount = 0, classes, data, recipient }: CustomDescP
...
+const BountyTxDescription = ({ classes, data }: CustomDescProps) => {
+  return (
+    <>
+      <Block className={classes.txData} data-testid={TRANSACTIONS_DESC_CUSTOM_DATA_TEST_ID}>
+        <Bold>Bounty payout:</Bold>
+        <div className={classes.txDataParagraph}>
+          {data.bountyLink && (
+            <p>
+              Bounty:{' '}
+              <a href={data.bountyLink} rel="noopener noreferrer" target="_blank">
```
Comment (Clearwood) [inline src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/index.jsx:262]: Please truncate the trailing zeroes of the Byte32 correctly :)

Comment (troggy) [reply src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/index.jsx:262]: there should be no trailing zeros already for new payouts. Once again, it applies only to new payouts — old ones created before the fix will still have zeroes and that is expected

Comment (Clearwood) [reply src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/index.jsx:262]: I am still experiencing the same problem with a now newly setup safe.

Comment (Clearwood) [reply src/routes/safe/components/Transactions/TxsTable/ExpandedTx/TxDescription/index.jsx:262]: I would leave this open, everything else works great and if we decide to use the Safe we could investigate this further on.

**Review by Clearwood — APPROVED** [review]

**Pull request merged**